### PR TITLE
Level, JobLevel & HeroLevel commands #305

### PR DESCRIPTION
**Pull request by Digitalis87**
Take care luci i'll take your job soon.

### CODE REVIEW

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 1]
```diff
@@ -305,6 +308,60 @@ public void Speed(SpeedPacket speedPacket)
             }
         }
 
+        [UsedImplicitly]
+        public void Level(SetLevelCommandPacket levelPacket)
+        {
+            Session.Character.Level = levelPacket.Level;
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: can we move the whole LevelSet thing to Character to avoid duplicate code later

Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: also if we can move this command to the WebApi stuff ;)

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 2]
```diff
@@ -305,6 +311,81 @@ public void Speed(SpeedPacket speedPacket)
...
+        [UsedImplicitly]
+        public void JobLevel(SetJobLevelCommandPacket jobLevelPacket)
+        {
+            if (jobLevelPacket.Name == null || jobLevelPacket.Name == Session.Character.Name)
+            {
+                Session.Character.SetJobLevel(jobLevelPacket.Level);
+                return;
+            }
+
+            var target = Broadcaster.Instance.GetCharacter(s => s.Name == jobLevelPacket.Name);
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: i would like this to be removed to only use the api on the commands.

File: src/NosCore.Core/Networking/WebApiAccess.cs [hunk 3]
```diff
@@ -209,35 +209,36 @@ public T Patch<T>(string route, object data, ServerConfiguration webApi)
             throw new HttpRequestException(postResponse.Headers.ToString());
         }
 
-        public void BroadcastPacket(PostedPacket postedPacket)
+        public void BroadcastPacket<T>(T packet, int channelId, string route)
```
Comment (erwan-joly) [inline src/NosCore.Core/Networking/WebApiAccess.cs]: this file doesn't need changes

File: src/NosCore.WorldServer/Controllers/StatController.cs [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            var session = Broadcaster.Instance.GetCharacter(s => s.Name == data.Character?.Name);
+
+            if (session == null)
+            {
+                return Ok(); //TODO: not found
+            }
+
+            switch (data.ActionType)
+            {
+                case UpdateStatActionType.UpdateLevel:
```
Comment (erwan-joly) [inline src/NosCore.WorldServer/Controllers/StatController.cs]: we need to send the packet for the refresh too

File: src/NosCore.GameObject/Character.cs [hunk 5]
```diff
@@ -355,7 +422,7 @@ public void SendRelationStatus(bool status)
                             Id = characterRelation.Value.RelatedCharacterId,
                             Name = characterRelation.Value.CharacterName
                         }
-                    });
+                    }, "api/packet");
```
Comment (erwan-joly) [inline src/NosCore.GameObject/Character.cs]: useless

File: src/NosCore.Controllers/DefaultPacketController.cs [hunk 6]
```diff
@@ -543,7 +543,7 @@ public void FriendTalk(BtkPacket btkPacket)
                 { Name = Session.Character.Name, Id = Session.Character.CharacterId },
                 OriginWorldId = MasterClientListSingleton.Instance.ChannelId,
                 ReceiverType = ReceiverType.OnlySomeone
-            }, receiver.ChannelId);
+            }, receiver.ChannelId, "api/packet");
```
Comment (erwan-joly) [inline src/NosCore.Controllers/DefaultPacketController.cs]: useless

File: src/NosCore.Controllers/DefaultPacketController.cs [hunk 7]
```diff
@@ -468,7 +468,7 @@ public void Whisper(WhisperPacket whisperPacket)
                     SenderCharacter = new Character { Name = Session.Character.Name },
                     OriginWorldId = MasterClientListSingleton.Instance.ChannelId,
                     ReceiverType = ReceiverType.OnlySomeone
-                }, receiver.ChannelId);
+                }, receiver.ChannelId, "api/packet");
```
Comment (erwan-joly) [inline src/NosCore.Controllers/DefaultPacketController.cs]: useless

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 8]
```diff
@@ -305,6 +311,81 @@ public void Speed(SpeedPacket speedPacket)
...
+
+            if (target != null)
+            {
+                target.SetJobLevel(jobLevelPacket.Level);
+                return;
+            }
+
+            WebApiAccess.Instance.BroadcastPacket(new StatData
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: this is definitly not a Packet use the normal ApiAccess Post

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 9]
```diff
@@ -305,6 +311,81 @@ public void Speed(SpeedPacket speedPacket)
...
+        [UsedImplicitly]
+        public void Level(SetLevelCommandPacket levelPacket)
+        {
+            if (levelPacket.Name == null || levelPacket.Name == Session.Character.Name)
+            {
+                Session.Character.SetLevel(levelPacket.Level);
+                return;
+            }
+
+            var target = Broadcaster.Instance.GetCharacter(s => s.Name == levelPacket.Name);
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: i would like this to be removed to only use the api on the commands.

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 10]
```diff
@@ -305,6 +311,81 @@ public void Speed(SpeedPacket speedPacket)
...
+        {
+            if (levelPacket.Name == null || levelPacket.Name == Session.Character.Name)
+            {
+                Session.Character.SetHeroLevel(levelPacket.Level);
+                return;
+            }
+
+            var target = Broadcaster.Instance.GetCharacter(s => s.Name == levelPacket.Name);
+
+            if (target != null)
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: i would like this to be removed to only use the api on the commands.

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 11]
```diff
@@ -170,7 +176,7 @@ public void Shout(ShoutPacket shoutPacket)
                 ReceiverType = ReceiverType.All
             };
 
-            WebApiAccess.Instance.BroadcastPackets(new List<PostedPacket>(new[] { sayPostedPacket, msgPostedPacket }));
+            WebApiAccess.Instance.BroadcastPackets(new List<PostedPacket>(new[] { sayPostedPacket, msgPostedPacket }), "api/packet");
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: useless

File: src/NosCore.WorldServer/Controllers/StatController.cs (hunk 4 above)
Comment (Elendan) [reply src/NosCore.WorldServer/Controllers/StatController.cs]: It's already in the methods that set the levels

File: src/NosCore.Core/Networking/WebApiAccess.cs [hunk 12]
```diff
@@ -233,7 +234,7 @@ public void BroadcastPackets(List<PostedPacket> packets)
             }
         }
 
-        public void BroadcastPackets(List<PostedPacket> packets, int channelId)
+        public void BroadcastPacket(List<PostedPacket> packets, int channelId)
```
Comment (erwan-joly) [inline src/NosCore.Core/Networking/WebApiAccess.cs]: why changing the name it take a list ><

File: src/NosCore.Data/WebApi/StatData.cs [hunk 13]
```diff
@@ -0,0 +1,35 @@
...
+
+namespace NosCore.Data.WebApi
+{
+    public class StatData
+    {
+        public Character Character { get; set; }
+
+        public UpdateStatActionType ActionType { get; set; }
+
+        public byte Data { get; set; }
```
Comment (erwan-joly) [inline src/NosCore.Data/WebApi/StatData.cs]: can be other type technicly but lets figure out this later (add a todo)

File: src/NosCore.GameObject/ComponentEntities/Interfaces/ICharacterEntity.cs [hunk 14]
```diff
@@ -71,5 +71,11 @@ public interface ICharacterEntity : INamedEntity, IExperiencedEntity
         ConcurrentDictionary<Guid, CharacterRelation> RelationWithCharacter { get; }
 
         ConcurrentDictionary<long, long> GroupRequestCharacterIds { get; }
+
+        void SetLevel(byte level);
```
Comment (erwan-joly) [inline src/NosCore.GameObject/ComponentEntities/Interfaces/ICharacterEntity.cs]: not link to Character ;) it's for the Interface that got a level (IExperiencedEntity maybe) in this case move the SetLevel to the extension (our pet can level up too)

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 15]
```diff
@@ -305,6 +311,144 @@ public void Speed(SpeedPacket speedPacket)
...
+                var target = accounts.FirstOrDefault(s => s.ConnectedCharacter.Name == levelPacket.Name);
+
+                if (target != null)
+                {
+                    receiver = target;
+                    config = channel.WebApi;
+                }
+            }
+
+            if (receiver == null)
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: todo remove this and use the Post 404 error

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 16]
```diff
@@ -305,6 +311,144 @@ public void Speed(SpeedPacket speedPacket)
...
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: todo remove this and 404 error

File: src/NosCore.Controllers/CommandPacketController.cs [hunk 17]
```diff
@@ -305,6 +311,144 @@ public void Speed(SpeedPacket speedPacket)
...
+                var target = accounts.FirstOrDefault(s => s.ConnectedCharacter.Name == levelPacket.Name);
+                
+                if (target != null)
+                {
+                    receiver = target;
+                    config = channel.WebApi;
+                }
+            }
+
+            if (receiver == null)
```
Comment (erwan-joly) [inline src/NosCore.Controllers/CommandPacketController.cs]: todo remove this and 404 post error

**Comment by erwan-joly** [comment]
nice work ;) some comments

**Pull request merged**